### ISSUE TITLE
Produce XML output file to publish detections via the IMOS portal

### ISSUE THREAD

**Issue by xhoenner** [aodn/aatams#249]
This enhancement would eventually allow people to filter and download acoustic telemetry detections straight from the IMOS portal, using an underlying WMS and WFS layer on Geoserver.

If I understand correctly, and @jkburges correct me if I'm wrong of feel free to expand this section, an output XML file can be produced from the webapp. Subsequently this XML could be routinely ingested by a Talend harvester that would massage the data to populate a _map and _data table for publication with Geoserver.

**Comment by jkburges** [comment]
I started thinking about this issue again due to https://github.com/aodn/aatams/pull/272, and I think the OP here is part of a broader issue.

The proposed solution in https://github.com/aodn/aatams/pull/272 puts what is essentially "business logic" (the calculation of surgery end dates) in to the Postgresql database.

This is sort of thing is already happening - particularly w.r.t. determining whether or not detections are valid.  In fact, AFAIK, we have logic spread across three areas:

* the DB
* the grails web-app
* the monthly reports (SQL, but not sure how they are managed)

There are a variety of reasons why this is so, among them, performance considerations and convenience.

However, we have to be careful with this fragmentation of logic: in my mind, it presents two main problems:

1. duplication of logic (and therefore doubled maintenance effort, divergence of behaviour etc)
1. security constraints are at particular risk of not being enforced

I think in an ideal world, the security constraints would be enforced at the database layer, for maximum safety. Secondly, all other non-security related business logic would be applied in a higher level app, an API if you will, as this is generally much more expressive and easy to understand, maintain, and if designed well, *should* also be easier to use (as someone writing the web-app or reports).

Everything else, i.e. what is currently the web-app presented to users, reports etc, would then feed of this.

**Comment by xhoenner** [comment]
Agreed, it'd be best if both the reports and the db tables used to create Geoserver layers could be generated without me having to design my own SQL queries. It wouldn't be trivial to implement though as the data reports collect information from ~ 25 different db tables.

There was much excitement at the AATAMS workshop when I demo'd a prototype of the future AATAMS layer (only public tags): https://portal-rc.aodn.org.au/imos123/home?uuid=0ede6b3d-8635-472f-b91c-56a758b4e091. @jkburges how many iterations do you think will be required to complete that task?

**Comment by jkburges** [comment]
> how many iterations do you think will be required to complete that task?

n + 1

**Comment by xhoenner** [comment]
@pblain the set of [QC scripts](https://github.com/aodn/aatams/tree/master/scripts/R/QC) that I have developed would also greatly benefit for security and accurateness sakes from a database extract based on the web app and database logic rather than an ad-hoc SQL query. This is particularly relevant if we were to run this QC algorithm not just as a one-off like now but thinking of an annual/bi-annual update.

The whole QC procedure is detailed [here](https://github.com/aodn/aatams/issues/283#issuecomment-274387863). In a nutshell raw valid detections, along with deployment metadata, for all registered tags are downloaded through a dump of the production db on the aatams-dev-instance machine. As a result of this operation one 'raw' file is produced per tag deployment. Those raw files are then independently processed to extract several metrics used for testing the likely validity of individual detections. 'Processed' files are thus produced, along with a 'TagMetadata.txt' file that holds all tag and tag deployment metadata. A third script then inspects those metadata, correcting a few inconsistencies, and deletes all 'Processed' files and records in 'TagMetadata.txt' for embargoed and protected tags. Note that this workflow was originally designed for the task team which wanted to use data for embargoed tags as it had obtained agreement from corresponding PIs to use their embargoed tag data and metadata in the context of a meta-analaysis.

**Comment by pblain** [comment]
The official position is that we shouldn't be bypassing security and running extracts against the backend database. I realise that it happened once, and I understand the reasons it was done, but I'm trusting that no more are done until we take stock and decide what to do about this predicament. 
I documented how we should implement the AATAMS extract process a few years ago. If more extracts like this are required we need to get it implemented in the webapp. Theoretically there should be a user story on the backlog for this. The pain point is, and as we're all acutely aware, we don't have enough developers to keep up with competing demands. An alternative idea, if you're keen ( @xhoenner ) is that we (the dev team) design it for you and get you to implement it in grails - with our feedback and support. Getting you involved on the grails side might add just enough man power to pull the AATAMS webapp out of its death spiral - maybe.

**Comment by pblain** [comment]
... actually "death spiral" isn't the right metaphor - it's more of a slow painful lingering death.
